File: typescript/src/pricing.ts
/**
 * Model pricing database for OpenAI and Anthropic.
 * Prices are in USD per 1 million tokens. Updated March 2026.
 */

export interface ModelPrice {
  inputPerMillion: number;
  outputPerMillion: number;
}

const MODEL_PRICES: Record<string, ModelPrice> = {
  // ── OpenAI ──────────────────────────────────────────────
  // GPT-4.1 family
  'gpt-4.1': { inputPerMillion: 2.0, outputPerMillion: 8.0 },
  'gpt-4.1-mini': { inputPerMillion: 0.4, outputPerMillion: 1.6 },
  'gpt-4.1-nano': { inputPerMillion: 0.1, outputPerMillion: 0.4 },
  // GPT-4o family
  'gpt-4o': { inputPerMillion: 2.5, outputPerMillion: 10.0 },
  'gpt-4o-mini': { inputPerMillion: 0.15, outputPerMillion: 0.6 },
  'gpt-4o-audio-preview': { inputPerMillion: 2.5, outputPerMillion: 10.0 },
  // GPT-4 legacy
  'gpt-4-turbo': { inputPerMillion: 10.0, outputPerMillion: 30.0 },
  'gpt-4': { inputPerMillion: 30.0, outputPerMillion: 60.0 },
  // GPT-3.5
  'gpt-3.5-turbo': { inputPerMillion: 0.5, outputPerMillion: 1.5 },
  // o-series reasoning
  'o3': { inputPerMillion: 2.0, outputPerMillion: 8.0 },
  'o3-mini': { inputPerMillion: 1.1, outputPerMillion: 4.4 },
  'o4-mini': { inputPerMillion: 1.1, outputPerMillion: 4.4 },
  'o1': { inputPerMillion: 15.0, outputPerMillion: 60.0 },
  'o1-mini': { inputPerMillion: 1.1, outputPerMillion: 4.4 },
  'o1-preview': { inputPerMillion: 15.0, outputPerMillion: 60.0 },
  // Embeddings
  'text-embedding-3-small': { inputPerMillion: 0.02, outputPerMillion: 0 },
  'text-embedding-3-large': { inputPerMillion: 0.13, outputPerMillion: 0 },
  'text-embedding-ada-002': { inputPerMillion: 0.1, outputPerMillion: 0 },

  // ── Anthropic ───────────────────────────────────────────
  'claude-opus-4-20250514': { inputPerMillion: 15.0, outputPerMillion: 75.0 },
  'claude-sonnet-4-20250514': { inputPerMillion: 3.0, outputPerMillion: 15.0 },
  'claude-3-7-sonnet-20250219': { inputPerMillion: 3.0, outputPerMillion: 15.0 },
  'claude-3-5-sonnet-20241022': { inputPerMillion: 3.0, outputPerMillion: 15.0 },
  'claude-3-5-sonnet-20240620': { inputPerMillion: 3.0, outputPerMillion: 15.0 },
  'claude-3-5-haiku-20241022': { inputPerMillion: 0.8, outputPerMillion: 4.0 },
  'claude-3-opus-20240229': { inputPerMillion: 15.0, outputPerMillion: 75.0 },
  'claude-3-sonnet-20240229': { inputPerMillion: 3.0, outputPerMillion: 15.0 },
  'claude-3-haiku-20240307': { inputPerMillion: 0.25, outputPerMillion: 1.25 },
  // Aliases
  'claude-opus-4-0': { inputPerMillion: 15.0, outputPerMillion: 75.0 },
  'claude-sonnet-4-0': { inputPerMillion: 3.0, outputPerMillion: 15.0 },
  'claude-3.7-sonnet': { inputPerMillion: 3.0, outputPerMillion: 15.0 },
  'claude-3.5-sonnet': { inputPerMillion: 3.0, outputPerMillion: 15.0 },
  'claude-3.5-haiku': { inputPerMillion: 0.8, outputPerMillion: 4.0 },
  'claude-3-opus': { inputPerMillion: 15.0, outputPerMillion: 75.0 },
  'claude-3-sonnet': { inputPerMillion: 3.0, outputPerMillion: 15.0 },
  'claude-3-haiku': { inputPerMillion: 0.25, outputPerMillion: 1.25 },
};

const ALIASES: Record<string, string> = {
  'gpt-4o-2024-11-20': 'gpt-4o',
  'gpt-4o-2024-08-06': 'gpt-4o',
  'gpt-4o-2024-05-13': 'gpt-4o',
  'gpt-4o-mini-2024-07-18': 'gpt-4o-mini',
  'gpt-4-turbo-2024-04-09': 'gpt-4-turbo',
  'gpt-4-turbo-preview': 'gpt-4-turbo',
  'gpt-4-0125-preview': 'gpt-4-turbo',
  'gpt-4-1106-preview': 'gpt-4-turbo',
  'gpt-3.5-turbo-0125': 'gpt-3.5-turbo',
  'gpt-3.5-turbo-1106': 'gpt-3.5-turbo',
  'o3-2025-04-16': 'o3',
  'o4-mini-2025-04-16': 'o4-mini',
};

export function resolveModel(model: string): string {
  return ALIASES[model] ?? model;
}

export function getPrice(model: string): ModelPrice | null {
  const canonical = resolveModel(model);
  return MODEL_PRICES[canonical] ?? null;
}

export function getCost(
  model: string,
  inputTokens: number,
  outputTokens: number
): number | null {
  const price = getPrice(model);
  if (!price) return null;
  return (
    (inputTokens * price.inputPerMillion +
      outputTokens * price.outputPerMillion) /
    1_000_000
  );
}

File: python/src/compute_cfo/pricing.py
"""Model pricing database for OpenAI and Anthropic.

Prices are in USD per 1 million tokens. Updated March 2026.
"""

from __future__ import annotations

from typing import Optional, Tuple

# {model_id: (input_price_per_1M, output_price_per_1M)}
MODEL_PRICES: dict[str, Tuple[float, float]] = {
    # ── OpenAI ──────────────────────────────────────────────
    # GPT-4.1 family
    "gpt-4.1": (2.00, 8.00),
    "gpt-4.1-mini": (0.40, 1.60),
    "gpt-4.1-nano": (0.10, 0.40),
    # GPT-4o family
    "gpt-4o": (2.50, 10.00),
    "gpt-4o-mini": (0.15, 0.60),
    "gpt-4o-audio-preview": (2.50, 10.00),
    # GPT-4 legacy
    "gpt-4-turbo": (10.00, 30.00),
    "gpt-4": (30.00, 60.00),
    # GPT-3.5
    "gpt-3.5-turbo": (0.50, 1.50),
    # o-series reasoning
    "o3": (2.00, 8.00),
    "o3-mini": (1.10, 4.40),
    "o4-mini": (1.10, 4.40),
    "o1": (15.00, 60.00),
    "o1-mini": (1.10, 4.40),
    "o1-preview": (15.00, 60.00),
    # Embeddings
    "text-embedding-3-small": (0.02, 0.0),
    "text-embedding-3-large": (0.13, 0.0),
    "text-embedding-ada-002": (0.10, 0.0),

    # ── Anthropic ───────────────────────────────────────────
    "claude-opus-4-20250514": (15.00, 75.00),
    "claude-sonnet-4-20250514": (3.00, 15.00),
    "claude-3-7-sonnet-20250219": (3.00, 15.00),
    "claude-3-5-sonnet-20241022": (3.00, 15.00),
    "claude-3-5-sonnet-20240620": (3.00, 15.00),
    "claude-3-5-haiku-20241022": (0.80, 4.00),
    "claude-3-opus-20240229": (15.00, 75.00),
    "claude-3-sonnet-20240229": (3.00, 15.00),
    "claude-3-haiku-20240307": (0.25, 1.25),
    # Aliases
    "claude-opus-4-0": (15.00, 75.00),
    "claude-sonnet-4-0": (3.00, 15.00),
    "claude-3.7-sonnet": (3.00, 15.00),
    "claude-3.5-sonnet": (3.00, 15.00),
    "claude-3.5-haiku": (0.80, 4.00),
    "claude-3-opus": (15.00, 75.00),
    "claude-3-sonnet": (3.00, 15.00),
    "claude-3-haiku": (0.25, 1.25),
}

# Common aliases mapping
_ALIASES: dict[str, str] = {
    "gpt-4o-2024-11-20": "gpt-4o",
    "gpt-4o-2024-08-06": "gpt-4o",
    "gpt-4o-2024-05-13": "gpt-4o",
    "gpt-4o-mini-2024-07-18": "gpt-4o-mini",
    "gpt-4-turbo-2024-04-09": "gpt-4-turbo",
    "gpt-4-turbo-preview": "gpt-4-turbo",
    "gpt-4-0125-preview": "gpt-4-turbo",
    "gpt-4-1106-preview": "gpt-4-turbo",
    "gpt-3.5-turbo-0125": "gpt-3.5-turbo",
    "gpt-3.5-turbo-1106": "gpt-3.5-turbo",
    "o3-2025-04-16": "o3",
    "o4-mini-2025-04-16": "o4-mini",
}


def resolve_model(model: str) -> str:
    """Resolve model aliases to canonical name."""
    return _ALIASES.get(model, model)


def get_price(model: str) -> Optional[Tuple[float, float]]:
    """Return (input_price, output_price) per 1M tokens, or None if unknown."""
    canonical = resolve_model(model)
    return MODEL_PRICES.get(canonical)


def get_cost(model: str, input_tokens: int, output_tokens: int) -> Optional[float]:
    """Calculate cost in USD for a given model and token counts.

    Returns None if model pricing is unknown.
    """
    price = get_price(model)
    if price is None:
        return None
    input_price, output_price = price
    return (input_tokens * input_price + output_tokens * output_price) / 1_000_000
